core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -654,31 +653,29 @@ private[zio] final class FiberContext[E, A](
     }
 
   @tailrec
-  private[this] final def enterAsync(): Boolean = {
+  private[this] final def enterAsync(ref: AnyRef): Boolean = {
     val oldState = state.get
 
     oldState match {
       case Executing(_, observers, interrupt) =>
-        val newState = Executing(FiberStatus.Suspended(interruptible), observers, interrupt)
+        val interruptible = this.interruptible
+        val newState      = Executing(FiberStatus.Suspended(interruptible, ref), observers, interrupt)
 
-        if (!state.compareAndSet(oldState, newState)) enterAsync()
-        else if (shouldInterrupt) {
-          // Fiber interrupted, so go back into running state:
-          exitAsync()
-          false
-        } else true
+        if (interruptible && interrupt) false
+        else if (!state.compareAndSet(oldState, newState)) enterAsync(ref)
+        else true
 
       case _ => false
     }
   }
 
   @tailrec
-  private[this] final def exitAsync(): Boolean = {
+  private[this] final def exitAsync(ref: AnyRef): Boolean = {
```

Comment:
We can type this more precisely than `AnyRef`; at least in your case it's `ZIO[_, _, _]`